ambry-store/src/test/java/com.github.ambry.store/CuratedLogIndexState.java

```diff
@@ -393,6 +474,119 @@ void verifyEntriesForHardDeletes(Set<MockId> deletedIds) throws IOException {
+    long expiresAtMs = allKeys.get(id).getFirst().getExpiresAtMs();
+    return expiresAtMs != Utils.Infinite_Time && expiresAtMs < referenceTimeMs;
+  }
+
+  /**
+   * Ensures that the {@link PersistentIndex} is sane and correct by checking that
+   * 1. It contains no duplicate entries.
+   * 2. The ordering of PUT and DELETE entries is correct.
+   * 3. There are no offsets in the log not accounted for in the index.
+   * @throws IOException
+   * @throws StoreException
+   */
+  void verifyRealIndexSanity() throws IOException, StoreException {
+    Map<MockId, Boolean> keyToDeleteSeenMap = new HashMap<>();
+    IndexSegment prevIndexSegment = null;
+    for (IndexSegment indexSegment : index.getIndexSegments().values()) {
+      Offset indexSegmentStartOffset = indexSegment.getStartOffset();
+      if (prevIndexSegment == null) {
+        assertEquals("There are offsets in the log not accounted for in index", log.getFirstSegment().getStartOffset(),
+            indexSegmentStartOffset.getOffset());
+      } else if (prevIndexSegment.getLogSegmentName().equals(indexSegment.getLogSegmentName())) {
+        assertEquals("There are offsets in the log not accounted for in index", prevIndexSegment.getEndOffset(),
+            indexSegmentStartOffset);
+      } else {
+        LogSegment segment = log.getSegment(prevIndexSegment.getLogSegmentName());
+        assertEquals("There are offsets in the log not accounted for in index", segment.getEndOffset(),
+            prevIndexSegment.getEndOffset().getOffset());
+        segment = log.getNextSegment(segment);
+        assertEquals("There are offsets in the log not accounted for in index", segment.getStartOffset(),
+            indexSegmentStartOffset.getOffset());
+      }
+      NavigableSet<IndexEntry> indexEntries = new TreeSet<>(PersistentIndex.INDEX_ENTRIES_COMPARATOR);
+      List<MessageInfo> infos = new ArrayList<>();
+      FindEntriesCondition condition = new FindEntriesCondition(Long.MAX_VALUE);
+      indexSegment.getEntriesSince(null, condition, infos, new AtomicLong(0));
+
+      for (MessageInfo info : infos) {
+        MockId id = (MockId) info.getStoreKey();
+        IndexValue value = indexSegment.find(id);
+        indexEntries.add(new IndexEntry(id, value));
+        Boolean deleteSeen = keyToDeleteSeenMap.get(id);
+        if (info.isDeleted()) {
+          if (deleteSeen != null) {
+            assertFalse("Duplicated DELETE record for " + id, deleteSeen);
+          }
+          keyToDeleteSeenMap.put(id, true);
+        } else {
+          if (deleteSeen != null) {
+            if (deleteSeen) {
+              fail("PUT record after delete record for " + id);
+            } else {
+              fail("Duplicated PUT record for " + id);
+            }
+          }
+          keyToDeleteSeenMap.put(id, false);
+        }
+      }
+      long expectedOffset = indexSegmentStartOffset.getOffset();
+      for (IndexEntry entry : indexEntries) {
```

Comment:
can you add a comment here stating that the "indexEntries" are ordered based on Offsets. I missed it when I started to review this (as index by default is ordered on Keys).